src/main/java/com/epam/aidial/core/limiter/RateLimiter.java

```diff
@@ -52,18 +76,26 @@ public RateLimitResult limit(ProxyContext context) {
             } else {
                 log.warn("Limit must be positive for deployment: {}", deployment.getName());
             }
-            return new RateLimitResult(HttpStatus.FORBIDDEN, "Access denied");
-        }
-
-        Id id = new Id(key.getKey(), deployment.getName());
-        RateLimit rate = rates.get(id);
-
-        if (rate == null) {
-            return RateLimitResult.SUCCESS;
+            return Future.succeededFuture(new RateLimitResult(HttpStatus.FORBIDDEN, "Access denied"));
         }
 
-        long timestamp = System.currentTimeMillis();
-        return rate.update(timestamp, limit);
+        String redisKey = getRedisKey(key.getKey(), deployment.getName());
+        return vertx.executeBlocking(() -> {
+            try (var ignore = lockService.lock(redisKey)) {
+                RBucket<String>  bucket = redis.getBucket(redisKey, StringCodec.INSTANCE);
+                String json = bucket.get();
+                RateLimit rateLimit;
+                if (json == null) {
+                    return RateLimitResult.SUCCESS;
+                } else {
+                    rateLimit = ProxyUtil.MAPPER.readValue(json, RateLimit.class);
+                }
+                long timestamp = System.currentTimeMillis();
+                RateLimitResult result = rateLimit.update(timestamp, limit);
+                bucket.set(ProxyUtil.MAPPER.writeValueAsString(rateLimit));
```

Comment:
Do you really need to set it back?

Reply:
We change the state of the rate limit(e.g. counts). So yes, we have to set it back.